extensions/amp-viewer-integration/0.1/amp-viewer-integration.js

```diff
@@ -56,31 +57,76 @@ export class AmpViewerIntegration {
     dev().fine(TAG, 'handshake init()');
     const viewer = viewerForDoc(this.win.document);
     this.unconfirmedViewerOrigin_ = viewer.getParam('origin');
-    if (!this.unconfirmedViewerOrigin_) {
-      dev().fine(TAG, 'Viewer origin not specified.');
-      return null;
+    this.unconfirmedViewerOrigin_ = ''; // mocking what would happen with a native app.
+
+    const win = this.win;
+    const unconfirmedViewerOrigin = this.unconfirmedViewerOrigin_;
+    if (this.unconfirmedViewerOrigin_) {
+      class WindowPortEmulator {
+        addEventListener(eventType, handler) {
+          listen(win, 'message', e => {
+            if (e.origin == unconfirmedViewerOrigin &&
+                e.source == win.parent && e.data.app == APP) {
+              handler(e);
+            }
+          });
+        }
+        postMessage(data) {
+          win.parent./*OK*/postMessage(data, unconfirmedViewerOrigin);
+        }
+      }
+      return this.openAndStart_(viewer, new WindowPortEmulator());
     }
 
-    dev().fine(TAG, 'listening for messages', this.unconfirmedViewerOrigin_);
-    const messaging = new Messaging(
-      this.win, this.win.parent, this.unconfirmedViewerOrigin_);
+    // port/webivew case
+    const preHandshake = new Promise(resolve => {
+      this.win.addEventListener('message', e => {
+        console.log('+++++++ampdoc got a message:', e.type, e.data);
+        // Viewer says: "I'm ready for you"
+        if (
+          // e.origin === '' && !e.source && //commenting out for now but need to uncomment before submit
+            e.data.app == APP &&
+            e.data.name == 'handshake-poll' &&
+            e.ports && e.ports.length == 1) {
+          resolve(e.ports[0]);
+        }
+      });
+    });
+    return preHandshake.then(port => {
+      class WindowPortEmulator {
```

Comment:
the problem is that messaging.js calles addEventListener rather than port.onmessage = handler.
If I removed this class I'd need to take care of 2 separate cases in messaging.js which is what we're trying to avoid